### PR TITLE
add an ERef object, like Haskell's IORef.

### PR DESCRIPTION
**Pull request by radix**
It only exposes `read` and `modify` operations for now, and doesn't implement the cool STM-style in-place updates, but the interface allows for it.

### CODE REVIEW

File: effect/ref.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+from ._base import Effect
+from ._dispatcher import TypeDispatcher
+from ._sync import sync_performer
+
+
+class ERef(object):
+    """
+    An effectful mutable variable.
+
+    Compare to Haskell's ``IORef`` or Clojure's ``atom``.
```
Comment (cyli) [inline effect/ref.py]: My only suggestion is to document the purpose of this object,  e.g. from an online discussion, that it's just a way to compose code purely, to indicate that a mutable variable is being changed.  Mostly because I was confused as to what purposes this served, since from reading atom's documentation it seems like `swap!` does something more than just changing a variable in place (I guess handles threading concurrency maybe?)

Not everyone is familiar with IORef or atom.  :)

File: effect/ref.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
+from characteristic import attributes
+
+from ._base import Effect
+from ._dispatcher import TypeDispatcher
+from ._sync import sync_performer
+
+
+class ERef(object):
```
Comment (glyph) [inline effect/ref.py]: Why "ERef"?  How about "MutableReference"?

Going with a naming convention because it mirrors the [famous beginner-friendliness](https://wiki.haskell.org/Monad_tutorials_timeline) of the Haskell community's standards or the [plainspoken clarity](https://glyph.twistedmatrix.com/2012/07/simple-made-variadic.html) of Clojure seems misguided :).

**Comment by glyph** [comment]
I can see why you would want something like `ModifyERef` - which I would call `SetAttribute`.  My understanding of the use-case goes something like this:

You have some code which transforms some structure into an Effect, or returns an Effect that expresses some intent around making an external service request.  But sometimes you want to talk to an external service, and sometimes you want to manipulate some internal state.  So it follows that while Python has a reasonably concise and efficient language around variable mutation, you'd still want a structure that could represent "variable mutation" as an Effect.

What I don't understand is what the point of `ERef` itself is.  It's an individual cell object which carries a reasonably high runtime cost (especially in CPython).  It's also a container for a *single* value, so it doesn't bridge nicely into code which has existing mutable structures in it.

If there were going to be a thing like `ERef`, it seems like it would be best to construct it something like `effect = MutableObject(object_with_attributes).attribute("attribute_name").set(new_value)`; that would make it easy to bridge an existing mutable-object interface.  Similarly, this conceptually bridges to `effect = MutableList(list_object).append(value)`, `effect = MutableDict(dict_object).set(key, value)`.

Perhaps my lack of understanding of `ERef`'s design hinges upon the "cool STM-style in-place updates" thing, which makes no sense to me.  STM seems like a thing the runtime has to do to get any benefit out of it (my understanding is that PyPy's STM would be happiest with a small number of naively defined un-shared structures).

File: effect/ref.py (hunk 1 above)
Comment (radix) [inline effect/ref.py]: I've updated this docstring a bit. Can you let me know if it makes more sense?

**Comment by radix** [comment]
@glyph

To clarify the "cool STM-style in-place updates": the simplest way to implement strong consistency in updates is by adding an atomic check-and-set on a version number associated with the ERef (in Python, we'd likely implement this with a lock, but maybe PyPy could expose some lower-level primitives).

And this leads to "why ModifyERef and not SetAttribute" -- It's `ModifyERef` and not `SetAttribute` because it takes a callable instead of a new value. Implementing it this way means that you can implement those atomic updates by repeatedly calling the transformation function if the version number mismatches, until you are able to write to the expected version and it succeeds.

So, ERef in that scenario would also carry a version number along with it.

Your `MutableObject` idea may have some merit, though I think it encourages a proliferation of mutable variables instead of isolating them to just what you need. And my current use case, in Otter, matches up with `ERef`.

Also: I don't care about performance on CPython. :-)

Ideally, what you put in an ERef is something that can be updated efficiently, like a persistent data structure that is implemented as a (H)AMT (e.g. something from the `pyrsistent` module). So you don't need to worry about actually mutating sub-components of your mutable object -- you just replace the whole object entirely, and do it efficiently, with something like a pyrsistent.PSet (or PRecord or PVec or whatever).

File: effect/ref.py (hunk 1 above)
Comment (cyli) [inline effect/ref.py]: It does make more sense, but I guess the main reason I was confused was I didn't understand what the difference between this and just mutating a variable was, so I thought I was missing something.

I'm not sure if this would fit better in this docstring or in a wiki, or howto, or some other documentation, but would it make sense to explain that this is basically that equivalent of python variable mutation, but how and why it's better?  The discussion below was fairly enlightening - I would have just had a callback that mutated some object or something, rather than use this effect with a transformer, and it might be useful to discuss why this pattern is better.

:)  Probably super obvious to you, less so to me.

Comment (radix) [inline effect/ref.py]: Alright, I'm going to try to explain it a bit more in-depth. I'm going to take a big step back and start from the beginning to try to explain how re-assigning a variable and making an HTTP request are the same kind of thing.

Effect has a few different reasons for existing, but each reason can stand on its own.

1. it separates performers from intents, so you can easily and naturally have a library that e.g. supports both blocking and non-blocking backends for an HTTP request.
2. intents are represented transparently so you can easily compare them in your tests
3. everything is purely functional, which has its own benefits for reasoning about code and for testability.

I'm mainly going to talk about the third one, though ERef also benefits from number 2.

It makes just as much sense to have an intent to represent an HTTP request as it does to have an intent to represent a mutable variable. Both of these things are side-effects, and both of them are otherwise trivial to do with typical Python -- you either call `request.get()` or you say `self.x = foo`. They both cause interactions with the outside world -- the request does network I/O, and the mutable variable assignment allows other logical threads of execution to observe some change in its environment.

Effect is a way to represent both of these scenarios with a consistent interface, where instead of making the HTTP request or reassigning the variable in your core logic, you represent those actions with an Effect, and the things that depend on those actions as callbacks to the Effect -- which are all truly pure functions.

So I think if someone is interested in representing their side-effects with Effect, to allow them to write their code purely functional or to make testing their code easy, it makes just as much sense for them to use Effect to represent mutable variables (in the cases where those mutable variables are truly necessary) as for IO.

I hope this helps a bit, and gives more context to this class. I'm happy to improve the ERef docstring more, but I think some of this probably does deserve to be in the more general Effect documentation. Is anything still unclear? If so, I'm very happy to continue discussing. If this explanation was clear enough, then can you offer any feedback on what might improve the ERef docstring?

Thanks for reviewing this and engaging me to come up with clearer descriptions for all this stuff :)

Comment (cyli) [inline effect/ref.py]: Very nice explanation!  Yes, this very clear!  I was just suggesting that this be documented somewhere, either in this docstring, or in the module docstring, or in a wiki or readme some place, because someone just coming across this class might wonder (as I did) why it existed.  The comment above I think can pretty much be copy-pasted - it's pretty awesome.

Thanks for taking the time to explain!

**Comment by glyph** [comment]
> To clarify the "cool STM-style in-place updates": the simplest way to implement strong consistency in updates is by adding an atomic check-and-set on a version number associated with the ERef (in Python, we'd likely implement this with a lock, but maybe PyPy could expose some lower-level primitives).

[PyPy's primitives for STM are "add to transaction" and "run transaction"](http://pypy.readthedocs.org/en/latest/stm.html#atomic-sections-transactions-etc-a-better-way-to-write-parallel-programs); you don't need to implement cells yourself.  The whole point of its STM layer is that most application code doesn't need to know that parallelism is happening.

> And this leads to "why ModifyERef and not SetAttribute" -- It's ModifyERef and not SetAttribute because it takes a callable instead of a new value. Implementing it this way means that you can implement those atomic updates by repeatedly calling the transformation function if the version number mismatches, until you are able to write to the expected version and it succeeds.

ModifyAttribute, then; I suppose I shouldn't have been talking about 

> So, ERef in that scenario would also carry a version number along with it.

> Your MutableObject idea may have some merit, though I think it encourages a proliferation of mutable variables instead of isolating them to just what you need. And my current use case, in Otter, matches up with ERef.

> Also: I don't care about performance on CPython. :-)

> Ideally, what you put in an ERef is something that can be updated efficiently, like a persistent data structure that is implemented as a (H)AMT (e.g. something from the pyrsistent module). So you don't need to worry about actually mutating sub-components of your mutable object -- you just replace the whole object entirely, and do it efficiently, with something like a pyrsistent.PSet (or PRecord or PVec or whatever).

**Comment by glyph** [comment]
> To clarify the "cool STM-style in-place updates": the simplest way to implement strong consistency in updates is by adding an atomic check-and-set on a version number associated with the ERef (in Python, we'd likely implement this with a lock, 

Do you mean "In CPython"?

> but maybe PyPy could expose some lower-level primitives).

[PyPy's primitives for STM are "add to transaction" and "run transaction"](http://pypy.readthedocs.org/en/latest/stm.html#atomic-sections-transactions-etc-a-better-way-to-write-parallel-programs); you don't need to implement revertable cells or write barriers yourself.  The whole point of its STM layer is that most application code doesn't need to know that parallelism is happening.

> And this leads to "why ModifyERef and not SetAttribute" -- It's ModifyERef and not SetAttribute because it takes a callable instead of a new value. Implementing it this way means that you can implement those atomic updates by repeatedly calling the transformation function if the version number mismatches, until you are able to write to the expected version and it succeeds.

`ModifyAttribute`, then; I suppose I shouldn't have been talking about a separate "set" in the context of an atomic-ish test-and-set.

> So, ERef in that scenario would also carry a version number along with it.

In what scenario?  The "with a lock" scenario or the "special new pypy primitives" scenario?  I browsed some docs on IORef and I didn't see anything related to version numbers or retries so I don't see how this would work.

It seems like this design is optimized for some parallel-universe future where `ERef` is somehow part of the PyPy runtime…

> Your MutableObject idea may have some merit, though I think it encourages a proliferation of mutable variables instead of isolating them to just what you need.

It's Python.  You already *have* plenty of mutable variables; they're everywhere. The stdlib is full of them, the runtime is full of them, they're not going away.  My contention here is that to facilitate adoption of Effect where it makes sense, ease of integration with those attributes is more important than an API which makes it more tedious to do so because rewriting all of that code in a different style would be more pure.

> And my current use case, in Otter, matches up with ERef.

It seems to line up just as well with a `MutableAttribute` though; just store the PSet directly on the instance where you're putting the `ERef` and then wrap a `MutableReference` around it.

> Also: I don't care about performance on CPython. :-)

Fair enough.  Otter is deployed on PyPy, I guess?

> Ideally, what you put in an ERef is something that can be updated efficiently, like a persistent data structure that is implemented as a (H)AMT (e.g. something from the pyrsistent module). So you don't need to worry about actually mutating sub-components of your mutable object -- you just replace the whole object entirely, and do it efficiently, with something like a pyrsistent.PSet (or PRecord or PVec or whatever).

If you're going to say "I don't care about performance" you can't turn around and say "efficiently" ;-).  pyrsistent is definitely not as efficient as just using a mutable built-in, especially in pypy, where mutable dicts are magical unicorns that have [CPU-cache-optimized algorithms](http://morepypy.blogspot.com/2015/01/faster-more-memory-efficient-and-more.html) :).  Even according to its own documentation, its performance ranges from "nearly on par" to "slower" than built-ins.

If you have a side-effect-ful cell though, using a plain set vs. using a PSet is an irrelevant implementation detail to an outside observer of your object; they still get an Effect which manipulates some internal state they shouldn't need to see.

**Comment by radix** [comment]
I had an in-person meeting with @glyph and @lvh (and also consulted @cyli and @manishtomar). I'll summarize some of the discussion points:

The most important point is that things that need to deal with "lots" of mutable variables should be wrapped in higher-level, application-specific intents. I've implemented a number of intents for Otter that are high level, like "Authenticate", "InvalidateAuthenticationToken", "ModifyGroupState", and so on. Another option is to use Func for ad-hoc effect-izing of legacy code, but usually defining a custom intent is the best option. I think this is a better option than having something like MutableObject/MutableDict/MutableList etc, since I think those would lead to too *many* effects, and perhaps even encourage more use of effect than necessary (not to mention that it could lead to an explosion of `Mutable*` classes).

We also discussed the STM stuff. The main point was: we could eventually implement this with low-level compare-and-set operation, in the same way that Java's `AtomicReference` works. This would even interoperate seamlessly with PyPy's STM. Doing a compare-and-set operation in a STM transaction would *not* be a conflict-generating operation, since by construction honors changes made by other threads. But this is super forward-looking and not really relevant now. Having an atomic compare-and-set is only necessary if you're modifying the same reference from multiple physical threads at the same time anyway. The current interface just gives you a simple way to derive new state from old state with a pure function. And we can add an modify_atomic that at least implements atomicity with a lock around the read/write until we get a proper compare-and-swap.

So basically I'm just going to rename `ERef` to `Reference` and improve the docstrings a bit.

**Comment by cyli** [comment]
As for the stated goal of changing the name and improving docstrings, LGTM!  Not sure if you want to get more input.

**Comment by radix** [comment]
@cyli - I think we've sorted out everything so I'm gonna merge. Thanks a lot for the help!
